src/ripple/app/paths/PathRequest.cpp

```diff
@@ -442,92 +441,141 @@ void PathRequest::resetLevel (int l)
         iLastLevel = l;
 }
 
-bool
-PathRequest::findPaths (
-    RippleLineCache::ref cache,
-        FindPaths& fp, Issue const& issue,
-            Json::Value& jvArray)
+std::unique_ptr<Pathfinder> const&
+PathRequest::getPathFinder(RippleLineCache::ref cache,
+    hash_map<Currency, std::unique_ptr<Pathfinder>>& currency_map,
+        Currency const& currency, STAmount const& dst_amount,
+            int const level)
 {
-    STPath fullLiquidityPath;
-    auto result = fp.findPathsForIssue(issue,
-        mContext[issue], fullLiquidityPath);
-    if (! result)
+    auto i = currency_map.find(currency);
+    if (i != currency_map.end())
+        return i->second;
+    auto pathfinder = std::make_unique<Pathfinder>(
+        cache, *raSrcAccount, *raDstAccount, currency,
+            boost::none, dst_amount, saSendMax);
+    if (pathfinder->findPaths(level))
+        pathfinder->computePathRanks(max_paths_);
+    else
+        pathfinder.reset();  // It's a bad request - clear it.
+    return currency_map[currency] = std::move(pathfinder);
+}
+
+void
+PathRequest::findPaths (RippleLineCache::ref cache, int const level,
+    Json::Value& jvArray)
+{
+    auto sourceCurrencies = sciSourceCurrencies;
+    if (sourceCurrencies.empty ())
     {
-        m_journal.debug << iIdentifier << " No paths found";
-        return false;
+        auto usCurrencies =
+            accountSourceCurrencies(*raSrcAccount, cache, true);
+        bool sameAccount = *raSrcAccount == *raDstAccount;
+        for (auto const& c : usCurrencies)
+        {
+            if (!sameAccount || (c != saDstAmount.getCurrency()))
+            {
+                if (c.isZero())
```

Comment:
I'd suggest this next as both shorter and clearer:

    auto& account = c.isZero() ? xrpAccount() : *raSrcAccount;
    sourceCurrencies.insert({c, account});